### PR TITLE
Add kudosu history to UserProfileOverlay

### PR DESCRIPTION
**Pull request by EVAST9919**
- [x] Depends on https://github.com/ppy/osu/pull/5858
Screenshot_1

### CODE REVIEW

**Comment by peppy** [comment]
```
[network:verbose] 2019-08-22 06:15:25: Performing request osu.Game.Online.API.Requests.GetUserKudosuHistoryRequest
[network:verbose] 2019-08-22 06:15:25: Request to https://osu.ppy.sh/api/v2/users/2/kudosu successfully completed!
[network:verbose] 2019-08-22 06:15:25: Processing response from https://osu.ppy.sh/api/v2/users/2/kudosu failed with Newtonsoft.Json.JsonSerializationException: Error setting value to 'action' on 'osu.Game.Online.API.Requests.Responses.APIKudosuHistory'. ---> System.ArgumentException: Requested value 'AllowKudosuGive' was not found.
```

you may want to go check the available types on `osu-web`.

File: osu.Game/Overlays/Profile/Sections/Kudosu/PaginatedKudosuHistoryContainer.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        protected override void ShowMore()
+        {
+            request = new GetUserKudosuHistoryRequest(User.Value.Id, VisiblePages++, ItemsPerPage);
+            request.Success += items => Schedule(() =>
+            {
+                MoreButton.FadeTo(items.Count == ItemsPerPage ? 1 : 0);
+                MoreButton.IsLoading = false;
+
+                if (!items.Any() && VisiblePages == 1)
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/Kudosu/PaginatedKudosuHistoryContainer.cs]: can't this copy-pasted logic be centralised?

```csharp
request.Success += UpdateItems(items.Select(i=> new DrawableKudosuHistoryItem(i));
```

then move all the rest of the logic (including the schedule) to UpdateItems method in `PaginatedContainer`.

File: osu.Game/Online/API/Requests/Responses/APIKudosuHistory.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            [JsonProperty("username")]
+            public string Username;
+        }
+
+        [JsonProperty("action")]
+        private string action
+        {
+            set
+            {
+                string parsed = value.Contains(".") ? value.Split('.')[0].Pascalize() + value.Split('.')[1].Pascalize() : value.Pascalize();
```
Comment (peppy) [inline osu.Game/Online/API/Requests/Responses/APIKudosuHistory.cs]: worth adding a comment saying what the input of this is.

File: osu.Game/Online/API/Requests/Responses/APIKudosuHistory.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+{
+    public class APIKudosuHistory
+    {
+        [JsonProperty("created_at")]
+        public DateTimeOffset CreatedAt;
+
+        [JsonProperty("amount")]
+        private int amount
+        {
+            set => Amount = Math.Abs(value);
```
Comment (peppy) [inline osu.Game/Online/API/Requests/Responses/APIKudosuHistory.cs]: add a comment about why we are absing

File: osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+                case KudosuAction.RestoreGive:
+                    linkFlowContainer.AddText(@"Received ");
+                    addKudosuPart();
+                    addMainPart(@" from modding post restoration of ");
+                    addPostPart();
+                    break;
+
+                case KudosuAction.RecalculateGive:
+                    linkFlowContainer.AddText(@"Received ");
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs]: these could be moved/grouped to a switch statement within `addKudosuPart` (which could them be called regardless of the action, returning in `default` if not requried).

File: osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+            string userLinkTemplate() => $"[{historyItem.Giver?.Url} {historyItem.Giver?.Username}]";
+
+            switch (historyItem.Action)
+            {
+                case KudosuAction.VoteGive:
+                    linkFlowContainer.AddText(@"Received ");
+                    addKudosuPart();
+                    addMainPart(@" from obtaining votes in modding post of ");
+                    addPostPart();
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs]: `addPostPart` is called in every case. move out of switch.

File: osu.Game/Overlays/Profile/Sections/PaginatedContainer.cs [hunk 6]
```diff
@@ -80,13 +84,57 @@ private void load(IAPIProvider api, RulesetStore rulesets)
 
         private void onUserChanged(ValueChangedEvent<User> e)
         {
+            loadCancellation?.Cancel();
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/PaginatedContainer.cs]: because this is changing a sizeable amount of important logic, please split this into its own PR.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/Profile/Sections/PaginatedContainer.cs (hunk 6 above)
Comment (EVAST9919) [reply osu.Game/Overlays/Profile/Sections/PaginatedContainer.cs]: Sure

File: osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+            switch (historyItem.Action)
+            {
+                case KudosuAction.VoteGive:
+                    return @" from obtaining votes in modding post of";
+
+                case KudosuAction.Give:
+                    return $@" from {userLink()} for a post at";
+
+                case KudosuAction.Reset:
+                    return $@"Kudosu reset by {userLink()} for the post";
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs]: needto be fixed ("Kudosu" should not be in string).

File: osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        private string getSource(APIKudosuHistory historyItem)
+        {
+            string userLink() => $"[{historyItem.Giver?.Url} {historyItem.Giver?.Username}]";
+
+            switch (historyItem.Action)
+            {
+                case KudosuAction.VoteGive:
+                    return @" from obtaining votes in modding post of";
```
Comment (peppy) [inline osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs]: leading space should be removed from all these items.

File: osu.Game/Online/API/Requests/Responses/APIKudosuHistory.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            [JsonProperty("username")]
+            public string Username;
+        }
+
+        [JsonProperty("action")]
+        private string action
+        {
+            set
+            {
+                //We will receive something like "foo.bar" or just "foo"
```
Comment (peppy) [inline osu.Game/Online/API/Requests/Responses/APIKudosuHistory.cs]: what is the "foo" and "bar"?

File: osu.Game/Online/API/Requests/Responses/APIKudosuHistory.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+{
+    public class APIKudosuHistory
+    {
+        [JsonProperty("created_at")]
+        public DateTimeOffset CreatedAt;
+
+        [JsonProperty("amount")]
+        private int amount
+        {
+            //We can receive negative values. However "action" is enough to build needed items
```
Comment (peppy) [inline osu.Game/Online/API/Requests/Responses/APIKudosuHistory.cs]: this comment doesn't really understand

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs (hunk 7 above)
Comment (EVAST9919) [reply osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs]: Why? Web has it like a string, for example

Comment (peppy) [reply osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs]: kudosu is included in the other part of the string?

Comment (EVAST9919) [reply osu.Game/Overlays/Profile/Sections/Kudosu/DrawableKudosuHistoryItem.cs]: Screenshot_1
This one is an exception

**Comment by peppy** [comment]
Rewrote this to match web implementation 1:1. Should have been done this way from the start.
